sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java

```diff
@@ -50,19 +50,44 @@
 @Evolving
 public abstract class ColumnVector implements AutoCloseable {
 
+  private int refCount = 1;
+
   /**
    * Returns the data type of this column vector.
    */
   public final DataType dataType() { return type; }
 
+  /**
+   * Increment the reference count for this vector. This is an implementation detail and
+   * only BoundReference should call this directly.
+   * @return this for easy chaining.
+   */
+  public final ColumnVector incRefCount() {
+    refCount++;
+    return this;
+  }
+
   /**
    * Cleans up memory for this column vector. The column vector is not usable after this.
+   * In reality it decrements the reference count and if it reaches 0 the resources are released
+   * but this is an implementation detail that most code should just ignore.
    *
    * This overwrites `AutoCloseable.close` to remove the `throws` clause, as column vector is
    * in-memory and we don't expect any exception to happen during closing.
    */
   @Override
-  public abstract void close();
+  public final void close() {
+    refCount--;
+    if (refCount == 0) {
+      doClose();
+    }
+  }
+
+  /**
+   * Actually cleans up memory for this column vector. The column vector is really not usable after
+   * this.
+   */
+  protected abstract void doClose();
```

Comment:
Can we use self-declarative name? This is because `columnVector` is public API for developers who want to support their storage.